### PR TITLE
Judge.me Coding Challenge - Full-stack Engineer

### PR DESCRIPTION
**Pull request by tatsuyalam**
1. Create new review page and implement background job. Rating radio reused available script.
2. Redesign products listing with reviews. Make reviews can be paginated via ajax call. **NOT implement** product pagination with AJAX for a better user experienced. Add sticky product photo, and 2 links: `Write a review` & `Shop info`
3. Create new shop page to display rating changes of of 2 consecutive months. Must run a new faker script to get distributed review months.

P/S: It's a long time for me to come back action views & ajax implementation. `Tailwindcss` is the first-time. BTW please help me to review and give me your feedbacks. Thanks!

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -58,3 +58,6 @@ end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+
+gem 'tailwindcss-rails', '~> 0.3.3'
```
Comment (damhonglinh) [inline Gemfile:62]: I've never hard of `tailwindcss` to be honest but it seems like a good choice.

File: README.md [hunk 2]
```diff
@@ -72,13 +80,18 @@ rails s
 
 # 10. Start Sidekiq in another terminal session
 sidekiq
+
+# 11. Start Webpack dev server
+bin/webpack-dev-server
```
Comment (damhonglinh) [inline README.md:85]: Thanks for adding the extra instruction :)

File: Gemfile [hunk 3]
```diff
@@ -58,3 +58,6 @@ end
...
+gem 'kaminari', '~> 1.2.1'
```
Comment (damhonglinh) [inline Gemfile:63]: Just FYI, there is a new pagination gem `pagy` (yes, another pagination gem) which is better (faster, simpler) than `kaminari`.

File: app/views/reviews/index.slim [hunk 4]
```diff
@@ -1,8 +1,32 @@
...
-    | Product #{datum[:product].title}
+.filter-shop.mx-4.flex.justify-end.items-center
+  span.mr-2 Filter by shop 
+  = select_tag :shop_select, options_for_select(@shops, @shop_id), { include_blank: "ALL SHOPS" }
 
-  - datum[:reviews].each do |review|
-    .rating
-      | Rating #{review.rating}
-    p.body == review.body
+- if @products.nil? || @products.length == 0
```
Comment (damhonglinh) [inline app/views/reviews/index.slim:5]: We usually use `.blank?` which I think can work here.

```suggestion
- if @products.blank?
```

File: app/views/reviews/new.slim [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          = f.text_field :reviewer_name, placeholder: "Enter your name", required: true
+      .row
+        .col 
+          = f.label :body, "Review"
+          = f.text_area :body, placeholder: "Write your review here", required: true, rows: 8
+      .row.flex.items-center
+        .col 
+          = f.submit "Submit Review"
+        .col.flex-1.text-right
+          = link_to "Back to list", reviews_path(shop_id: @product.shop_id, anchor: "p#{@product.id}")
```
Comment (damhonglinh) [inline app/views/reviews/new.slim:46]: Nice touch! I can see you went beyond simply doing the challenge, by making the features more finished, more usable.

File: app/javascript/stylesheets/application.scss [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  }
```
Comment (damhonglinh) [inline app/javascript/stylesheets/application.scss:59]: Very interesting way of writing the star rating form

File: app/controllers/reviews_controller.rb [hunk 7]
```diff
@@ -1,38 +1,65 @@
...
+    @review = Review.new
+  end
 
-    tags = tags_with_default(params)
-    Review.create(product_id: params[:product_id], body: params[:body], rating: params[:rating], reviewer_name: params[:reviewer_name], tags: tags)
+  # POST /
+  def create
+    product = Product.find_by(id: review_params[:product_id])
+    if product
+      tags = tags_with_default(review_params, product)
```
Comment (damhonglinh) [inline app/controllers/reviews_controller.rb:36]: It's probably to move this into the Worker or a service object than keeping it here in the controller.

File: app/controllers/reviews_controller.rb [hunk 8]
```diff
@@ -1,38 +1,65 @@
...
+      # Create reviews in background. No need to show errors (if any) to users, it's fine to skip creating the review silently when some validations fail.
+      ReviewSubmitWorker.perform_async(review_params.merge(tags: tags).to_h)
```
Comment (damhonglinh) [inline app/controllers/reviews_controller.rb:38]: Nitpick: I don't think we need to call `.to_h` as Sidekiq/Rails can do it automatically already.

File: Gemfile (hunk 1 above)
Comment (tatsuyalam) [reply Gemfile:62]: `tailwindcss` is trending guy beyond `bootstrap`

File: app/javascript/packs/shop.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// Register months period selection changes
+var selection = document.getElementById("months_select");
+if (selection) {
+  selection.addEventListener("change", (e) => {
+    var href = new URL(window.location.href);
+    href.searchParams.set("months", e.target.value);
+    // Redirect to new shop
```
Comment (damhonglinh) [inline app/javascript/packs/shop.js:7]: Nitpick: the comment seems irrelevantly copied from `app/javascript/packs/reviews.js`.

File: app/javascript/stylesheets/application.scss (hunk 6 above)
Comment (tatsuyalam) [reply app/javascript/stylesheets/application.scss:59]: Key point is `direction: rtl;` of the wrapper

File: app/controllers/reviews_controller.rb [hunk 10]
```diff
@@ -1,38 +1,65 @@
...
-      offset = params[:page].to_i * params[:per_page]
-
-      @data = []
-      products = Product.where("shop_id = #{params[:shop_id]}").sort_by(&:created_at)[offset..(offset + params[:per_page])]
-      products.each do |product|
-        reviews = product.reviews.sort_by(&:created_at)[offset..(offset + params[:per_page])]
-        @data << { product: product, reviews: reviews }
-      end
+    # Get store listing for filter
+    @shops = Shop.order(:name).map { |s| [s.name, s.id] }
```
Comment (damhonglinh) [inline app/controllers/reviews_controller.rb:7]: This would be simpler and faster for DB and the web server:
```suggestion
    @shops = Shop.order(:name).pluck(:name, :id)
```

File: app/controllers/reviews_controller.rb (hunk 7 above)
Comment (tatsuyalam) [reply app/controllers/reviews_controller.rb:36]: Yeap. I think I can reuse product object here to reduce a db query. BTW this logic should not place here (maybe in a callback and another worker)

File: app/controllers/reviews_controller.rb [hunk 11]
```diff
@@ -1,38 +1,65 @@
...
-  def create
-    # TODO: Create reviews in background. No need to show errors (if any) to users, it's fine to skip creating the review silently when some validations fail.
+  # GET /of_product
+  def of_product
+    @product_id = params[:product_id]
+    @init = params[:init] # trigger to scroll
+    if @product_id
+      list = Review.where(product_id: @product_id)
+      @total_reviews = list.count
+      @reviews = list.offset(@offset).limit(@per_page)
```
Comment (damhonglinh) [inline app/controllers/reviews_controller.rb:22]: Looks like this is missing `order(created_at: :desc)`.

File: app/views/reviews/_product_reviews.slim [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+          em.text-gray-400 Anonymous
+      .rating
+        span.text-lg.text-yellow-400.mr-2
+          - current_rate = review.rating.to_i
+          - current_rate.times do |r|                
+            span ★
+          - (5 - current_rate).times do |r|
+            span ☆
+        span.text-sm.text-gray-400 = time_ago_in_words(review.created_at)
+      p == review.body
```
Comment (damhonglinh) [inline app/views/reviews/_product_reviews.slim:24]: This is one of the hidden easter eggs. This is due to a convenient feature of Rails but is very dangerous (for security) if misused. 
https://stackoverflow.com/a/27173999/1815779

`==` means Rails won't escape HTML when printing the variable (`review.body` in this case) to view. And `review.body` is user generated content, so it can contain malicious JS (e.g. `<script>alert(1)</script>`). 

But you didn't spot this probably because you haven't worked on Rails views for a while.

File: app/views/reviews/of_product.js.slim [hunk 13]
```diff
@@ -0,0 +1,10 @@
+|
+  var product = document.querySelector("#p#{@product_id}");
+  var reviews = product.querySelector(".reviews");
+  // replace content
+  reviews.innerHTML = "#{ j render("product_reviews", product_id: @product_id, reviews: @reviews, total_reviews: @total_reviews, page: @page) }";
+
+// scroll to top
+- if @init.nil?
+  |
+    product.scrollIntoView();
```
Comment (damhonglinh) [inline app/views/reviews/of_product.js.slim:10]: It's not a good practice to return `js` format because it would couple browser and server more. 

It's better for the server to just return data (ideally in JSON, but can be in HTML for quicker implementation), and then browser (or client side) can use its own JS to handle the responded data.

File: app/javascript/packs/reviews.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+    // Redirect to new shop
+    window.location.href = href.toString();
+  });
+}
+
+// Fetch reviews in loading
+var products = document.querySelectorAll("[data-product-id]");
+products.forEach((e) => {
+  var link = e.querySelector(".load-reviews");
+  link.click();
```
Comment (damhonglinh) [inline app/javascript/packs/reviews.js:16]: Quite related to https://github.com/tatsuyalam/test-product-reviews/pull/1/files#r664225698

It's better to have the JS here to make the AJAX request, and append the responded data to the page (rather than having a remote link), and then in here, we can call that JS directly rather than calling `.click()`.

**Comment by damhonglinh** [comment]
@tatsuyalam Thanks for submitting the code!

Your code looks great in general, despite a few issues. Here is a brief assessment:

# Good points (many but we will just list some):
- Go beyond just doing the challenge, by making the features more finished, more usable
- Tick, qualify almost every points in our `Expectations` (in README)
- Know some Rails tricks that we believe only experienced RoR developers can know

# Not-so-good points (we will list exhaustively all the points):
- We can't see the use of service, query objects (probably due to time constraint)
- We can't see the use of BEM convention when naming CSS classes (probably due to using `Tailwindcss` lib)
- We can't see RSpec tests. But it's is our bad that we didn't include it in README. Our apologies!
- A few minor coding mistakes (e.g. copying irrelevant comment, missing `order(created_at: :desc)`)
- A few performance issues (e.g. `.map` vs `.pluck`). Performance is important for us, but (if you join us) we can learn these things a long the way. 
- A few debatable/opinionated styles (e.g. JS response format vs JSON/HTML response format), but (if you join us) we can discuss and sort these things out easily. 


Please note that we list all the `Not-so-good points` exhaustively, but only list a few example good points. So in general, your code looks great!

File: app/views/reviews/of_product.js.slim (hunk 13 above)
Comment (tatsuyalam) [reply app/views/reviews/of_product.js.slim:10]: Yes. JSON will be better. But I want to implement as the original way of rail.

**Comment by tatsuyalam** [comment]
Thanks for your time. And I hope we can do together @judge.me
